runtime/vam/expr/coerce.go

```diff
@@ -36,235 +34,7 @@ func coerceVals(sctx *super.Context, a, b vector.Any) (vector.Any, vector.Any, v
-	switch id := val.Type().ID(); {
-	case super.IsUnsigned(id):
-		return float64(val.Uint()), true
-	case super.IsSigned(id):
-		return float64(val.Int()), true
-	case super.IsFloat(id):
-		return val.Float(), true
-	case id == super.IDString:
-		v, err := byteconv.ParseFloat64(val.Bytes())
-		return v, err == nil
-	}
-	return 0, false
-}
-
-func ToUint(val super.Value) (uint64, bool) {
-	switch id := val.Type().ID(); {
-	case super.IsUnsigned(id):
-		return val.Uint(), true
-	case super.IsSigned(id):
-		v := val.Int()
-		if v < 0 {
-			return 0, false
-		}
-		return uint64(v), true
-	case super.IsFloat(id):
-		return uint64(val.Float()), true
-	case id == super.IDString:
-		v, err := byteconv.ParseUint64(val.Bytes())
-		return v, err == nil
-	}
-	return 0, false
-}
-
-func ToInt(val super.Value) (int64, bool) {
-	switch id := val.Type().ID(); {
-	case super.IsUnsigned(id):
-		return int64(val.Uint()), true
-	case super.IsSigned(id):
-		// XXX check if negative? should -1:uint64 be maxint64 or an error?
-		return val.Int(), true
-	case super.IsFloat(id):
-		return int64(val.Float()), true
-	case id == super.IDString:
-		v, err := byteconv.ParseInt64(val.Bytes())
-		return v, err == nil
-	}
-	return 0, false
-}
-
-func ToBool(val super.Value) (bool, bool) {
-	if val.IsString() {
-		v, err := byteconv.ParseBool(val.Bytes())
-		return v, err == nil
-	}
-	v, ok := ToInt(val)
-	return v != 0, ok
-}
-
-// XXX should be numToFloat and handle time, dur, eventually decimal
-// (though we should promote float to decimal)
-func intToFloat(val vector.Any) vector.Any {
-	switch val := val.(type) {
-	case *vector.Int:
-		vals := val.Values
-		n := int(val.Len())
-		f := make([]float64, n)
-		for k := 0; k < n; k++ {
-			f[k] = float64(vals[k])
-		}
-		return vector.NewFloat(super.TypeFloat64, f, val.Nulls)
-	case *vector.Uint:
-		vals := val.Values
-		n := int(len(vals))
-		f := make([]float64, n)
-		for k := 0; k < n; k++ {
-			f[k] = float64(vals[k])
-		}
-		return vector.NewFloat(super.TypeFloat64, f, val.Nulls)
-	case *vector.Const:
-		f, ok := ToFloat(val.Value())
-		if !ok {
-			panic("ToFloat failed")
-		}
-		return vector.NewConst(super.NewFloat64(f), val.Len(), val.Nulls)
-	case *vector.Dict:
-		return vector.NewDict(intToFloat(val.Any), val.Index, val.Counts, val.Nulls)
-	case *vector.View:
-		return vector.Pick(intToFloat(val.Any), val.Index)
-	default:
-		panic(fmt.Sprintf("intToFloat invalid type: %T", val))
-	}
-}
-
-// XXX need intToUint (e.g, compare int to time?)
-func uintToInt(val vector.Any) vector.Any {
-	switch val := val.(type) {
-	case *vector.Uint:
-		vals := val.Values
-		n := int(len(vals))
-		out := make([]int64, n)
-		for k := 0; k < n; k++ {
-			out[k] = int64(vals[k])
-		}
-		return vector.NewInt(super.TypeInt64, out, val.Nulls)
-	default:
-		panic(fmt.Sprintf("intToFloat invalid type: %T", val))
-	}
+	id, _ := coerce.Promote(super.NewValue(a.Type(), nil), super.NewValue(b.Type(), nil))
+	typ, _ := super.LookupPrimitiveByID(id)
```

Comment:
I know neither of these is expected to fail, but if that changes, ignoring these errors will probably make someone's life harder. 
```suggestion
	id, err := coerce.Promote(super.NewValue(a.Type(), nil), super.NewValue(b.Type(), nil))
	if err != nil {
		panic(err)
	}
	typ, err := super.LookupPrimitiveByID(id)
		if err != nil {
		panic(err)
	}
```

Reply:
yeah good point